### PR TITLE
Add fake content and test build Craft+Gatsby

### PR DESCRIPTION
**Pull request by mattstein**
### Description

My goal is to take this project and use GitHub Actions to spin up the Craft site and run a Gatsby build using its content. This should make it easier to merge Snyk’s endless PRs with confidence and offer rudimentary smoke testing of the Craft + Gatsby integration.

It finally works, though I’d like to make sure I’m not doing anything wildly dull. The greatest potential offense is my forceful 777 permissions on Craft’s writable directories, which I can’t seem to back down to a reasonable level successfully:

https://github.com/craftcms/starter-blog/blob/574b9ceac22f30533a40bb3d0b9efd855d811214/.github/workflows/test.yml#L74-L75

The [GitHub Action yaml](https://github.com/craftcms/starter-blog/compare/feature/content-migration?expand=1#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88) should otherwise speak for itself.

- Get Craft instance running with PHP 7.4 CLI (with MySQL and PostgreSQL just to be paranoid).
- Run Craft content migration so there’s stuff to query and a known/consistent GraphQL token.
- Add nginx so Gatsby has a GraphQL URL to query for its build. (`http://nginx/api`)
- Step into Gatsby project, add Gatsby CLI locally, and build the static site using [gatsby-source-craft](https://github.com/craftcms/gatsby-source-craft).
- If the Gatsby build fails, upload Craft’s logs as build artifacts.

The only not-obvious struggle was that the GitHub Actions virtual environment [is *not* a Docker container itself](https://github.community/t/services-name-resolution-error/16716/5)—though you can specify one—so it doesn’t get networking benefits like host aliases. This is why I cheated by duplicating Docker aliases to `/etc/hosts`:

https://github.com/craftcms/starter-blog/blob/574b9ceac22f30533a40bb3d0b9efd855d811214/.github/workflows/test.yml#L65-L66

Is there anything that should be improved here, for safety or simplicity or just dignity?

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+jobs:
+  test:
+    name: Gatsby + Craft with PHP ${{ matrix.php-versions }} + ${{ matrix.db }}
+    runs-on: ${{ matrix.operating-system }}
+    strategy:
+      matrix:
+        operating-system: [ubuntu-latest]
+        php-versions: ["7.4"]
+        db: ["mysql", "pgsql"]
+    services:
```
Comment (jasonmccallister) [inline .github/workflows/test.yml:20]: @mattstein the spacing here might make more sense visually to add a new line before the "section" (e.g. a new line before "services" and make the rest of the file consistent - the steps are not grouped.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        run: cp .github/workflows/.env.${{ matrix.db }} .env
+
+      - name: Validate composer.json and composer.lock
+        run: composer validate
+
+      - name: Install Composer dependencies
+        run: composer install --prefer-dist --no-progress
+      
+      - name: Add Docker database host aliases to build environment
+        run: sudo echo "127.0.0.1 nginx postgres mysql" | sudo tee -a /etc/hosts
```
Comment (jasonmccallister) [inline .github/workflows/test.yml:67]: @mattstein I'm not sure if this is required... but I'm also not sure if the actions are in the same network... docker typically puts containers in the same network so you can reach them by name without any changes.

File: .github/workflows/test.yml (hunk 1 above)
Comment (mattstein) [reply .github/workflows/test.yml:21]: Done.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: Ping GraphQL endpoint
+        run: |-
+          curl -s -X POST -H "Content-Type: application/json" --data '{ "query": "{ ping }" }' http://nginx/api
+
+      - name: Prepare Gatsby environment
+        run: cp .github/workflows/.env-gatsby headless-front-end/gatsby/.env
+
+      - name: Install Gatsby dependencies
+        run: node --version && cd headless-front-end/gatsby && npm install && npm install gatsby-cli
```
Comment (jasonmccallister) [inline .github/workflows/test.yml:86]: @mattstein do you need the version output here or was this for debugging purposes?

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+      - name: Build Gatsby site
+        run: cd headless-front-end/gatsby && npm run build
```
Comment (jasonmccallister) [inline .github/workflows/test.yml:89]: I'm pretty sure you can set the working directory underneath the run command.

`working-directory: ./headless-front-end/gatsby`

File: .github/workflows/test.yml (hunk 3 above)
Comment (jasonmccallister) [reply .github/workflows/test.yml:86]: pretty sure you can just set the `working-directory` argument here.

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+      - name: Upload Craft logs
+        uses: actions/upload-artifact@v2
+        if: failure()
```
Comment (jasonmccallister) [inline .github/workflows/test.yml:93]: first time I have seen this used and that is amazing! 💯

**Review by jasonmccallister — CHANGES_REQUESTED** [review]
Looks amazing but I have a few questions

File: .github/workflows/test.yml (hunk 2 above)
Comment (mattstein) [reply .github/workflows/test.yml:67]: This approach may not be @jasonmccallister, but it’s absolutely essential the way things are set up. I mentioned that in the PR description and linked to the source that helped me realize it.

The GitHub Actions runner isn’t a Docker container so it doesn’t get the networking perks—hence this inelegant workaround.

File: .github/workflows/test.yml (hunk 3 above)
Comment (mattstein) [reply .github/workflows/test.yml]: I sure don’t need that, good catch @jasonmccallister. Used `working-directory` argument and it looks much nicer—didn’t realize I could do that.

**Review by jasonmccallister — APPROVED** [review]

**Pull request merged**